// create-ts-index/dist/CTIUtility.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CTIUtility {
    static addDot(ext) {
        if (ext.startsWith('.'))
            return ext;
        return `.${ext}`;
    }
    static addNewline(option, data) {
        if (option.addNewline)
            return `${data}\n`;
        return data;
    }
    static isNotEmpty(value) {
        return value !== undefined && value !== null;
    }
    static isEmpty(value) {
        return CTIUtility.isNotEmpty(value);
    }
    static parseBool(value) {
        if (value === undefined) {
            return false;
        }
        if (value === null) {
            return false;
        }
        // prevent 0 to false
        if (typeof value === 'number') {
            return true;
        }
        // literal 'false' to false boolean type
        if (typeof value === 'string' && value === 'false') {
            return false;
        }
        if (typeof value === 'string' && value === 'true') {
            return true;
        }
        return Boolean(value);
    }
    static getQuote(value) {
        if (value === 'd' || value === '"') {
            return '"';
        }
        return "'";
    }
}
exports.CTIUtility = CTIUtility;
